src/lib/trace.c

```diff
@@ -69,494 +70,110 @@ static void put_header(volatile uint32_t *dst, uint64_t timestamp)
-		- sizeof(uint32_t) * message_size_dwords)
-		trace->pos = 0;
-
-	spin_unlock_irq(&trace->lock, flags);
-
-	put_header(t, time);
-	t[payload_offset] = log_entry;
-	t[payload_offset + 1] = param1;
-	t[payload_offset + 2] = param2;
-	t[payload_offset + 3] = param3;
-
-	/* writeback trace data */
-	dcache_writeback_region((void *)t,
-		sizeof(uint32_t) * message_size_dwords);
-}
-
-void _trace_event_mbox_atomic3(uint32_t log_entry, uint32_t param1,
-	uint32_t param2, uint32_t param3)
-{
-	volatile uint32_t *t;
-	uint32_t message_size_dwords = MESSAGE_SIZE(3);
-	uint32_t payload_offset = sizeof(struct log_entry_header)
-		/ sizeof(uint32_t);
-	uint32_t dt[MESSAGE_SIZE(3)];
-	uint64_t time;
-
-	if (!trace->enable)
-		return;
-
-	time = platform_timer_get(platform_timer);
-
-	put_header(dt, time);
-	dt[payload_offset] = log_entry;
-	dt[payload_offset + 1] = param1;
-	dt[payload_offset + 2] = param2;
-	dt[payload_offset + 3] = param3;
-	dtrace_event_atomic((const char *)dt,
-		sizeof(uint32_t) * message_size_dwords);
-
-	/* write timestamp and event to trace buffer */
-	t = (volatile uint32_t *)(MAILBOX_TRACE_BASE + trace->pos);
-	trace->pos += sizeof(uint32_t) * message_size_dwords;
-
-	if (trace->pos > MAILBOX_TRACE_SIZE
-		- sizeof(uint32_t) * message_size_dwords)
-		trace->pos = 0;
-
-	put_header(t, time);
-	t[payload_offset] = log_entry;
-	t[payload_offset + 1] = param1;
-	t[payload_offset + 2] = param2;
-	t[payload_offset + 3] = param3;
-
-	/* writeback trace data */
-	dcache_writeback_region((void *)t,
-		sizeof(uint32_t) * message_size_dwords);
-}
+// TODO: it should be possible to for arbitrary array, passed as argument
+#define _TRACE_EVENT_NTH_IMPL_DT_STEP(i, _) dt[payload_offset + 1 + i] = CONCAT(param, i);
+#define _TRACE_EVENT_NTH_IMPL_T_STEP(i, _)   t[payload_offset + 1 + i] = CONCAT(param, i);
+#define _TRACE_EVENT_NTH_IMPL_DT(arg_count) SEQ_FROM_0_TO(arg_count, _TRACE_EVENT_NTH_IMPL_DT_STEP)
+#define _TRACE_EVENT_NTH_IMPL_T(arg_count)  SEQ_FROM_0_TO(arg_count, _TRACE_EVENT_NTH_IMPL_T_STEP)
+
+/* _trace_event function with poor people's version of constexpr if */
+#define _TRACE_EVENT_NTH_IMPL(is_mbox, is_atomic, arg_count)\
+_TRACE_EVENT_NTH(CONCAT(\
+IF_ELSE(is_mbox)(_mbox)(),   /*if is_mbox  , append _mbox   to function name*/\
+IF_ELSE(is_atomic)(_atomic)()/*if is_atomic, append _atomic to function name*/\
+), arg_count)\
+{\
+	uint32_t MESSAGE_SIZE_dwords = MESSAGE_SIZE(arg_count);\
+	uint32_t payload_offset = sizeof(struct log_entry_header)\
+		/ sizeof(uint32_t);\
+	uint32_t dt[MESSAGE_SIZE_dwords];\
+	IF_ELSE(is_mbox)\
+	(\
+		volatile uint32_t *t;\
+		IF_ELSE(is_atomic)()(unsigned long flags;)\
+	)()\
+\
+	if (!trace->enable)\
+		return;\
+\
+	IF_ELSE(is_mbox)\
+	(\
+		uint64_t time = platform_timer_get(platform_timer);\
+		put_header(dt, time);\
+	)\
+	(\
+		put_header(dt, platform_timer_get(platform_timer));\
+	)\
+\
+	dt[payload_offset] = log_entry;\
+	IF_ELSE(arg_count)(_TRACE_EVENT_NTH_IMPL_DT(arg_count))()\
```

Comment:
what's the benefit here of IF_ELSE ? I know the pre-processor will remove unused code but the compiler will do the same since these variables will be constants.

Reply:
@lgirdwood 
So with:

`IF_ELSE(is_atomic)()(unsigned long flags;)`
compiler warns about unused variable 'flags'.

```
IF_ELSE(is_mbox)\
(\
	uint64_t time = platform_timer_get(platform_timer);\
	put_header(dt, time);\
)\
(\
	put_header(dt, platform_timer_get(platform_timer));\
)\
```
is self explanatory.

`IF_ELSE(arg_count)(_TRACE_EVENT_NTH_IMPL_DT(arg_count))()`
is in fact a bit unnecessary, macro
`_TRACE_EVENT_NTH_IMPL_DT(arg_count)`
works just as well.

And why not something like `void (const is_mbox, const is_atomic, ...)`?
While this works
```
IF_ELSE(is_mbox)\
(\
	volatile uint32_t *t;\
	IF_ELSE(is_atomic)()(unsigned long flags;)\
)()\
```
something like
```
if(is_mbox)
{
	volatile uint32_t *t;
	if(is_atomic) { unsigned long flags; }
}
```
does in fact get removed with `is_mbox`/`is_atomic` being `false` (what we want), but when its `true` then those variables are valid only in those `if` blocks (what we don't want).

The only way the second could work would be
```
void _trace_event_main (const is_mbox, const is_atomic, ...) {
	if (is_mbox) {
		if (is_atomic) _trace_event_mbox_atomic(...);
		else _trace_event_mbox(...);
	} else {
		if (is_atomic) _trace_event_atomic(...);
		else _trace_event(...);
	}
}
void _trace_event_mbox_atomic(...){ ... }
void _trace_event_atomic(...){ ... }
void _trace_event_mbox(...){ ... }
void _trace_event(...){ ... }
```

then we get rid macro and compiler **should** optimise (based on constnes of `is_mbox` and `is_atomic`) all those `if`s in `_trace_event_main` into
```
_trace_event_main(1,1,...) {_trace_event_mbox_atomic(...)}
_trace_event_main(0,1,...) {_trace_event_atomic(...)}
_trace_event_main(1,0,...) {_trace_event_mbox(...)}
_trace_event_main(0,0,...) {_trace_event(...)}
```
but we also lose all information about which *code is shared between those functions*.

If the case was that those functions bear very little resemblance, then I'd in fact propose such approach.
But here, for instance `_trace_eventN` and `_trace_event_atomicN` differ only by 1 function call (but both are with same arguments).

As a human, I have to read and compare `_trace_event0` with `_trace_event_atomic0`, then  `_trace_event1` with `_trace_event_atomic1`, ..., then  `_trace_eventN` with `_trace_event_atomicN` to know how exactly they behave and notice the fact that they differ only in this one specific place. Reading only couple of them could make me unaware of for instance that if N is divisible by 3 then some additional logic applies.

It is not undoable sure, and once I read it once after each change made to them I know how they work, but in my opinion this introduces unnecessary complexity and points of failure. Entire point of DRY principle is to avoid such a situation when we update foo1, foo2, foo4, but miss foo3. This also lengthens the review process I imagine when you see for instance not X changes in 1 function (total X), but X changes in N places (total X*N).

Going further, `_trace_event_mbox0` is basically `_trace_event0` with additional block of code at the end,
and `_trace_event_mbox_atomic0` introduces to`_trace_event_mbox0` same changes as `_trace_event_atomic0` did to `_trace_event0` plus 2 additional function calls.
Sure, they well could be implemented as 4 separate blocks of code, but I don't think they should be.

I too don't like the kind of syntax using macro haxx introduce here, and I very much wish for something like if-constexpr in C++17, so it would look something like
```
void _trace_event_main (const is_mbox, const is_atomic, ...) {
	const uint32_t msg_size_dwords = MESSAGE_SIZE(arg_count);
	const uint32_t payload_offset = sizeof(struct log_entry_header) sizeof(uint32_t);
	uint32_t dt[MESSAGE_SIZE(arg_count)];
	if constexpr (is_mbox) {
		volatile uint32_t *t;
		if constexpr(is_atomic) { unsigned long flags; }
	}
	...
	if constexpr (is_mbox) {
	}
```
And let the compiler, not much less capable preprocessor, do the dirty work.

But we C here, so this is not a possibility.

The only part here that is **REALLY** confusing and hideous and that I maybe should comment better is this part
```
_TRACE_EVENT_NTH(CONCAT(\
IF_ELSE(is_mbox)(_mbox)(),   /*if is_mbox  , append _mbox   to function name*/\
IF_ELSE(is_atomic)(_atomic)()/*if is_atomic, append _atomic to function name*/\
), arg_count)\
{\
```

which gets expanded (depending on arg_count, is_mbox and is_atomic) into
```
void _trace_event0 (uint32_t log_entry ) {
void _trace_event_mbox0 (uint32_t log_entry ) {
void _trace_event_atomic0 (uint32_t log_entry ) {
void _trace_event_mbox_atomic0 (uint32_t log_entry ) {
...
void _trace_event3 (uint32_t log_entry , uint32_t param0 , uint32_t param1 , uint32_t param2) {
void _trace_event_mbox3 (uint32_t log_entry , uint32_t param0 , uint32_t param1 , uint32_t param2) {
void _trace_event_atomic3 (uint32_t log_entry , uint32_t param0 , uint32_t param1 , uint32_t param2) {
void _trace_event_mbox_atomic3 (uint32_t log_entry , uint32_t param0 , uint32_t param1 , uint32_t param2) {
```

but once again, `_trace_event_mbox_atomic3` differs with `_trace_event_mbox_atomic1` in exactly 3 places
`void _trace_event3 (uint32_t log_entry , uint32_t param0 , uint32_t param1 , uint32_t param2) {`
instead of
`void _trace_event1 (uint32_t log_entry , uint32_t param0) {`
then
```
	dt[payload_offset] = log_entry;
	dt[payload_offset + 1] = param1;
	dt[payload_offset + 2] = param2;
	dt[payload_offset + 3] = param3;
```
instead of
```
	dt[payload_offset] = log_entry;
	dt[payload_offset + 1] = param1;
```
and
```
	t[payload_offset] = log_entry;
	t[payload_offset + 1] = param1;
	t[payload_offset + 2] = param2;
	t[payload_offset + 3] = param3;
```
instead of
```
	t[payload_offset] = log_entry;
	t[payload_offset + 1] = param1;
```

All those 3 places could be described in pseudocode / high-level as `do foo(i) for every [0,N>`, so we get
```
void _trace_event3 (uint32_t log_entry , size_t param_count {, uint32_t param(i) } for every [0,param_count>)`
...
	dt[payload_offset] = log_entry;
	{ dt[payload_offset + 1 + i] = param(i); } for every} for every [0,param_count>
...
	t[payload_offset] = log_entry;
	{ t[payload_offset + 1 + i] = param(i); } for every} for every [0,param_count>
```
A little bit shorter, maybe more straightforward. Keep in mind, we could theoretically have _trace_event10 and then there would be even greater savings.

If to get to this conclusion (that this piece of code only depends on param_count) I have to read **16** functions (4 `_trace event` functions * 4 combinations of params) instead of `1`, then I think something's not right. And by read I de facto mean *run far too many diffs*.
Moreover, if we at some point decide that there should be `_trace_event5`, then we land with `20` total, for `6` `24`, and so on.

In my opinion introducing `IF_ELSE` (both conceptually simple and easy to use when its already written) to highlight to anyone that will maintain this code that *listen dude, here are two methods that differ only in this one particular single place depending on this specific argument* outweighs the level of obfuscation that is introduced. I would of course not recommend using it in places where only few LOCs are shared.

Regarding looping/recursion done in preprocessor, especially knowing how they work and what are their limitations in preprocessor, it is a tough one.
On one hand I understand you that those are inelegant haxx with their limitations and that what I tried to achieve should be really done in compiler, even just because compiler knows **SO MUCH** more about the source code than the preprocessor.

On the other hand though, there are only 2 other options. Either have separate method for each `_trace_event` parameter combination (what we had before), resulting in `param_count` times more code, or use proper vararg functions, as in https://en.cppreference.com/w/c/variadic.

But from what I read, their optimisation differs greatly between compilers, especially x86_64, and introduce some overhead.
I am aware that `premature optimisation is the root of all evil`, but from what I understand we want as little overhead for debugability as possible, to always keep up the traces production with the rate of events to be logged happening. And this is where vararg macros seem to have advantage over vararg functions, in that macros are 100% sure to be evaluated `not in run-time`, whereas with vararg functions we don't have that certainty.
Not even mentioning the fact that variadic functions do not provide any type safety. With macros, we still should get warning if we try to pass a `float` or `uint64_t` to `_trace_event0(uint32_t log_entry)`.

This is not 100% perfect comparison, but showcases the overhead of C variadic functions due to the need of argument parsing : http://quick-bench.com/EDRuG-zSTHFGB_DnHmIOajgVxf8

There is yet another option, of doing this in **C++**, but I highly doubt such option would even be considered :)

If in your opinion benefits do not outweigh the costs, then we can scrap this. But in my opinion it would be harmful in the long run. If you know of a better way of being more DRY here, that I maybe hadn't thought of, then I'd be happy to kill these macro haxx with fire and do it more elegant way.

Reply:
ok, thanks for the write up ! - golden rule is it must be simple and compile into minimal code. Its really simple for the programmer to manually select trace1(), trace2 etc. This code really boils down to a  variable size (know at compile time) memcpy with locking.

Reply:
Regarding minimal compile size
When I run following command in repo root
```
(\
  styleAll(){ awk -v style="$1" -v regex="$2" '{gsub(regex,"\033["style"m&\033[0m");print}';};\
  findFunc(){ awk -v func_regexp="$1" -v counter=-1 -v func_found=0 '
{
  if (func_found == 0) {func_found = match($0, func_regexp)}
  if (func_found != 0) {
    if (counter ==  0 ) {counter = -1; func_found =  0}
    split($0, chars, "")
    for (i=1; i <= length($0); i++) {
      if (chars[i]=="}") {--counter}
      if (chars[i]=="{") {if (counter<0) { counter = 0 }++counter}
    }
    if (func_found) { printf("%s\n", $0) }
  }
}';};\
  gcc -I src/platform/cannonlake/include -I src/arch/host/include -I src/include -E -dD src/lib/trace.c 2>/dev/null\
  |findFunc '^[ ]*void[ ]*_trace_event'\
  |sed\
    -e 's/;\ /;\n\t/g'\
    -e 's/{\ /\n{\n\t/g'\
    -e 's/}[\ ]*/\n}\n/g'\
  |styleAll 33 '[{}()\\[\\],;]'\
  |styleAll "1;31" '->|\\.|&'\
  |styleAll "1;35" '\\s(if|while|do|for|return|exit|panic)'\
  |styleAll "1;36" '(u?int(32|64)_t|void|struct|const|char|volatile|static|unsigned|long)\\s*\\**'\
  |styleAll "32" '0x[A-Za-z0-9]+'\
  |styleAll "7;32" 'param[0-9]+'\
  |less -r\
)
```
I get prettily printed code after preproc and before compile (using `gcc`).
When I run it in both master and this PR, I get identical code, minus whitespace changes and things like `1+2` instead of `3` (what will be optimised by compiler anyway).
So yeah, this solution does not cost any *compile size* (it does though cost *compile time*).

Reply:
Regarding simplicity
This solution is conceptually simple, but implementation is indeed not simple.
One have to know a thing or two about preprocessor to understand implementation (less to actually use).